rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp

```diff
@@ -0,0 +1,359 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_RCLCPP_ALLOCATOR_MEMORY_STRATEGY_HPP_
+#define RCLCPP_RCLCPP_ALLOCATOR_MEMORY_STRATEGY_HPP_
+
+#include <memory>
+#include <vector>
+
+#include <rclcpp/allocator/allocator_common.hpp>
+#include <rclcpp/memory_strategy.hpp>
+#include <rclcpp/node.hpp>
+
+namespace rclcpp
+{
+
+namespace memory_strategies
+{
+
+namespace allocator_memory_strategy
+{
+
+/// Delegate for handling memory allocations while the Executor is executing.
+/**
+ * By default, the memory strategy dynamically allocates memory for structures that come in from
+ * the rmw implementation after the executor waits for work, based on the number of entities that
+ * come through.
+ */
+template<typename Alloc = std::allocator<void>>
+class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
+{
+
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(AllocatorMemoryStrategy<Alloc>);
+
+  using ExecAllocTraits = allocator::AllocRebind<executor::AnyExecutable, Alloc>;
+  using ExecAlloc = typename ExecAllocTraits::allocator_type;
+  using ExecDeleter = allocator::Deleter<ExecAlloc, executor::AnyExecutable>;
+  using VoidAllocTraits = typename allocator::AllocRebind<void *, Alloc>;
+  using VoidAlloc = typename VoidAllocTraits::allocator_type;
+
+  using WeakNodeVector = std::vector<std::weak_ptr<node::Node>>;
+
+  AllocatorMemoryStrategy(std::shared_ptr<Alloc> allocator)
+  {
+    executable_allocator_ = std::make_shared<ExecAlloc>(*allocator.get());
+    allocator_ = std::make_shared<VoidAlloc>(*allocator.get());
+  }
+
+  AllocatorMemoryStrategy()
+  {
+    executable_allocator_ = std::make_shared<ExecAlloc>();
+    allocator_ = std::make_shared<VoidAlloc>();
+  }
+
+  size_t fill_subscriber_handles(void ** & ptr)
+  {
+    size_t max_size = subscriptions_.size();
+    subscriber_handles_.resize(max_size);
+    size_t count = 0;
+    for (auto & subscription : subscriptions_) {
+      subscriber_handles_[count++] = subscription->get_subscription_handle()->data;
+      if (subscription->get_intra_process_subscription_handle()) {
+        subscriber_handles_[count++] =
+          subscription->get_intra_process_subscription_handle()->data;
+      }
+    }
+    ptr = subscriber_handles_.data();
+    return count;
+  }
```

Comment:
We also talked about this offline, but I think these functions sort of don't belong in a memory strategy class. I agree that they probably don't belong in the Executor class either, so maybe they need another container class which is responsible for handling data within it. The executor can make requests of the container and the container can delegate allocation to the memory strategy or an allocator where appropriate.

I don't think we have to solve this design problem in this pull request, but we should consider doing this refactoring soon. Perhaps I'll do it in the second part of rclcpp library work, after the alpha 2 release.